yapapi/storage/gftp.py

```diff
@@ -243,22 +244,28 @@ async def upload_stream(self, length: int, stream: AsyncIterator[bytes]) -> Sour
         with open(file_name, "wb") as f:
             async for chunk in stream:
                 f.write(chunk)
-        process = await self.__get_process()
-        links = await process.publish(files=[str(file_name)])
-        assert len(links) == 1, "invalid gftp publish response"
-        link = links[0]
-        return GftpSource(length, link)
+        return await self.upload_file(file_name)
 
     async def upload_file(self, path: os.PathLike) -> Source:
+        hasher = hashlib.sha3_256()
+        with open(path, "rb") as f:
+            while True:
+                bytes = f.read(4096)
+                if not bytes:
+                    break
+                hasher.update(bytes)
+        digest = hasher.hexdigest()
+        if digest in self._registered_sources:
```

Comment:
Originally, the keys in `self._registered_sources` were file paths, which prevented errors resulting from publishing the same file twice. But errors were printed when publishing a file with the same contents as a file already published, in particular, when publishing random files resulting from `send_json()` when recomputing a failed task.